From dbd7e03c4af5ee4b462bcc238f49c35bbf43e8b8 Mon Sep 17 00:00:00 2001
From: Eric Timmons <etimmons@mit.edu>
Date: Sun, 2 Dec 2018 14:20:24 -0500
Subject: [PATCH 3/3] Fix foreign tests on musl libc

Musl libc's implementation of dlclose(3) is (intentionally) a noop (see:
https://wiki.musl-libc.org/functional-differences-from-glibc.html). Add a build
time test for this and a new feature :os-dlclose-is-noop. Additionally, use this
feature to skip some regression tests if it is present.
---
 tests/foreign.test.sh                         | 42 +++++++++++--------
 tools-for-build/Makefile                      |  3 ++
 tools-for-build/grovel-features.sh            |  7 ++++
 .../os-dlclose-is-noop-test-helper.c          |  1 +
 tools-for-build/os-dlclose-is-noop-test.c     | 19 +++++++++
 5 files changed, 55 insertions(+), 17 deletions(-)
 create mode 100644 tools-for-build/os-dlclose-is-noop-test-helper.c
 create mode 100644 tools-for-build/os-dlclose-is-noop-test.c

diff --git a/tests/foreign.test.sh b/tests/foreign.test.sh
index 7fb757813..78f0517a0 100755
--- a/tests/foreign.test.sh
+++ b/tests/foreign.test.sh
@@ -248,16 +248,20 @@ cat > $TEST_FILESTEM.test.lisp <<EOF
   (assert (= 13 foo))
   (assert (= 42 (bar)))
   (note "/original definitions ok")
-  (rename-file "$TEST_FILESTEM-b.so" "$TEST_FILESTEM-b.bak")
-  (rename-file "$TEST_FILESTEM-b2.so" "$TEST_FILESTEM-b.so")
-  (load-shared-object (truename "$TEST_FILESTEM-b.so"))
-  (note "/reloading ok")
-  (assert (= 42 foo))
-  (assert (= 13 (bar)))
-  (note "/redefined versions ok")
-  (rename-file "$TEST_FILESTEM-b.so" "$TEST_FILESTEM-b2.so")
-  (rename-file "$TEST_FILESTEM-b.bak" "$TEST_FILESTEM-b.so")
-  (note "/renamed back to originals")
+  #+os-dlclose-is-noop
+  (note "/skipping reloading tests")
+  #-os-dlclose-is-noop
+  (progn
+    (rename-file "$TEST_FILESTEM-b.so" "$TEST_FILESTEM-b.bak")
+    (rename-file "$TEST_FILESTEM-b2.so" "$TEST_FILESTEM-b.so")
+    (load-shared-object (truename "$TEST_FILESTEM-b.so"))
+    (note "/reloading ok")
+    (assert (= 42 foo))
+    (assert (= 13 (bar)))
+    (note "/redefined versions ok")
+    (rename-file "$TEST_FILESTEM-b.so" "$TEST_FILESTEM-b2.so")
+    (rename-file "$TEST_FILESTEM-b.bak" "$TEST_FILESTEM-b.so")
+    (note "/renamed back to originals"))
 
   ;; test late resolution
   #+linkage-table
@@ -274,13 +278,17 @@ cat > $TEST_FILESTEM.test.lisp <<EOF
     (load-shared-object (truename "$TEST_FILESTEM-c.so"))
     (assert (= 43 late-foo))
     (assert (= 14 (late-bar)))
-    (unload-shared-object (truename "$TEST_FILESTEM-c.so"))
-    (multiple-value-bind (val err) (ignore-errors late-foo)
-      (assert (not val))
-      (assert (typep err 'undefined-alien-error)))
-    (multiple-value-bind (val err) (ignore-errors (late-bar))
-      (assert (not val))
-      (assert (typep err 'undefined-alien-error)))
+    #+os-dlclose-is-noop
+    (note "/skipping unloading tests")
+    #-os-dlclose-is-noop
+    (progn
+      (unload-shared-object (truename "$TEST_FILESTEM-c.so"))
+      (multiple-value-bind (val err) (ignore-errors late-foo)
+        (assert (not val))
+        (assert (typep err 'undefined-alien-error)))
+      (multiple-value-bind (val err) (ignore-errors (late-bar))
+        (assert (not val))
+        (assert (typep err 'undefined-alien-error))))
     (note "/linkage table ok"))
 
   (sb-ext:exit :code $EXIT_LISP_WIN) ; success convention for Lisp program
diff --git a/tools-for-build/Makefile b/tools-for-build/Makefile
index 3f6e4ecf9..731ae6640 100644
--- a/tools-for-build/Makefile
+++ b/tools-for-build/Makefile
@@ -16,6 +16,9 @@ LDLIBS:=$(OS_LIBS)
 
 all: grovel-headers determine-endianness where-is-mcontext
 
+os-dlclose-is-noop-test-helper.so: os-dlclose-is-noop-test-helper.c
+	@$(CC) $(LDFLAGS) -shared $< -o $@ $(LOADLIBES) $(LDLIBS)
+
 clean:
 	rm -f *.o grovel-headers determine-endianness where-is-mcontext
 	rm -f *.exe
diff --git a/tools-for-build/grovel-features.sh b/tools-for-build/grovel-features.sh
index 9992b46ad..4e2546878 100644
--- a/tools-for-build/grovel-features.sh
+++ b/tools-for-build/grovel-features.sh
@@ -36,3 +36,10 @@ featurep os-provides-poll
 featurep os-provides-netdb-internal
 
 featurep arm-softfp
+
+# We need a helper shared library to test os-dlclose-is-noop
+$GNUMAKE os-dlclose-is-noop-test-helper.so > /dev/null 2>&1
+
+featurep os-dlclose-is-noop
+
+rm -f os-dlclose-is-noop-test-helper.so
diff --git a/tools-for-build/os-dlclose-is-noop-test-helper.c b/tools-for-build/os-dlclose-is-noop-test-helper.c
new file mode 100644
index 000000000..4be7a8e5b
--- /dev/null
+++ b/tools-for-build/os-dlclose-is-noop-test-helper.c
@@ -0,0 +1 @@
+int sbcl_dl_close_test = 42;
diff --git a/tools-for-build/os-dlclose-is-noop-test.c b/tools-for-build/os-dlclose-is-noop-test.c
new file mode 100644
index 000000000..3678870d5
--- /dev/null
+++ b/tools-for-build/os-dlclose-is-noop-test.c
@@ -0,0 +1,19 @@
+/* test to build and run so that we know if we have a noop dlclose
+ */
+
+#include <dlfcn.h>
+#include <stddef.h>
+
+int main ()
+{
+   void * handle = dlopen("./os-dlclose-is-noop-test-helper.so", RTLD_NOW | RTLD_GLOBAL);
+   dlclose(handle);
+
+   handle = dlopen("./os-dlclose-is-noop-test-helper.so", RTLD_NOW | RTLD_NOLOAD);
+
+   if (handle != NULL) {
+       return 104;
+   } else {
+       return 0;
+   }
+}
-- 
2.19.2

